From 5a50dfad800d991225ae1a637a29232bf887a790 Mon Sep 17 00:00:00 2001
From: Ray Donnelly <mingw.android@gmail.com>
Date: Sat, 29 Oct 2016 17:43:00 +0100
Subject: [PATCH 25/42] Win32 (MSVC): Add WinMain() for rstudio executable

JOM builds need WinMain()
Visual Studio builds need main()

.. and I am not sure why.
---
 src/cpp/desktop/DesktopMain.cpp | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/src/cpp/desktop/DesktopMain.cpp b/src/cpp/desktop/DesktopMain.cpp
index e0e35fd..46b6df3 100644
--- a/src/cpp/desktop/DesktopMain.cpp
+++ b/src/cpp/desktop/DesktopMain.cpp
@@ -42,6 +42,10 @@
 #include "DesktopSessionLauncher.hpp"
 #include "DesktopProgressActivator.hpp"
 
+#if defined(_MSC_VER)
+#include <TCHAR.h>
+#endif
+
 QProcess* pRSessionProcess;
 QString sharedSecret;
 
@@ -195,6 +199,7 @@ bool isNonProjectFilename(QString filename)
 
 } // anonymous namespace
 
+
 int main(int argc, char* argv[])
 {
    core::system::initHook();
@@ -382,3 +387,15 @@ int main(int argc, char* argv[])
    }
    CATCH_UNEXPECTED_EXCEPTION
 }
+
+#ifdef _WIN32
+// When building with jom, WinMain must be used, when building with Visual Studio, main must be used.
+// I am not sure why, but it is something I need to fix.
+int WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPSTR lpCmdLine, int nCmdShow)
+{
+	int argc = __argc;
+	char **argv = __argv;
+	main(argc, argv);
+}
+#endif
+
-- 
2.10.1

